cpp/src/hash/fixed_hash_set.cuh

```diff
@@ -0,0 +1,132 @@
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <hash/helper_functions.cuh>
+
+namespace cudf {
+
+/*
+ *  fixed set on a device
+ */
+template <typename Element, typename Hasher = default_hash<Element>,
+          typename Equality = equal_to<Element>>
+class device_fixed_hash_set {
+public:
+  device_fixed_hash_set(size_type hash_size, const size_type *hash_begin, const Element *hash_data)
+    : hash_size{hash_size}, hash_begin{hash_begin}, hash_data{hash_data}, hasher(), equals() {
+  }
+
+  bool __device__ contains(Element e) const {
+    size_type loc = hasher(e) % (2 * hash_size);
+
+    for (size_type i = hash_begin[loc] ; i < hash_begin[loc+1] ; ++i) {
+      if (equals(hash_data[i],e))
+        return true;
+    }
+
+    return false;
+  }
+
+private:
+  Hasher           hasher;
+  Equality         equals;
+  size_type        hash_size;
+  const size_type *hash_begin;
+  const Element   *hash_data;
+};
+
+/*
+ * Fixed size set on a device.  
+ *
+ */
+template <typename Element, typename Hasher = default_hash<Element>,
+          typename Equality = equal_to<Element>>
+class fixed_hash_set {
+public:
+  /**---------------------------------------------------------------------------*
+   * @brief Factory to construct a new fixed_device_set
+   *---------------------------------------------------------------------------**/
+  static fixed_hash_set<Element> create(column_view const& col,
+                                        cudaStream_t stream) {
+
+    auto d_column = column_device_view::create(col);
+    auto d_col = *d_column;
+
+    rmm::device_vector<size_type> hash_bins_start(2 * d_col.size() + 1, size_type{0});
+    rmm::device_vector<size_type> hash_bins_end(2 * d_col.size() + 1, size_type{0});
+    rmm::device_vector<Element>   hash_data(d_col.size());
+
+    Hasher hasher;
+    size_type *d_hash_bins_start = thrust::raw_pointer_cast(hash_bins_start.data());
+    size_type *d_hash_bins_end = thrust::raw_pointer_cast(hash_bins_end.data());
+    Element *d_hash_data = thrust::raw_pointer_cast(hash_data.data());
+
+    thrust::for_each(rmm::exec_policy(stream)->on(stream),
+                     thrust::make_counting_iterator<size_type>(0),
+                     thrust::make_counting_iterator<size_type>(col.size()),
+                     [d_hash_bins_start, d_col, hasher] __device__ (size_t idx) {
+                       if (!d_col.is_null(idx)) {
+                         Element e = d_col.element<Element>(idx);
+                         size_type tmp = hasher(e) % (2 * d_col.size());
+                         atomicAdd(d_hash_bins_start + tmp, size_type{1});
+                       }
+                     });
+
+    thrust::exclusive_scan(rmm::exec_policy(stream)->on(stream),
+                           hash_bins_start.begin(),
+                           hash_bins_start.end(),
+                           hash_bins_end.begin());
+
+    thrust::copy(rmm::exec_policy(stream)->on(stream),
```

Comment:
Is this copy needed? `thrust::scan` can be done in place.

Reply:
I need two copies of the array.  The hash_bins_start copy gets moved at the end of the factory function into the fixed_hash_set object as the base offset for each hash bin.  The hash_bins_end copy gets corrupted during the subsequent foreach call.  So even if I did the in place thrust::scan I would need to make a copy.